# BGC Donor Data Consolidation and Analysis

## Project Overview
The purpose of this project was to consolidate donor data from multiple
source systems into a single source of truth and perform donor analytics.

## Project Steps
1. Load and clean individual source datasets
2. Consolidate datasets into a master donation dataset
3. Review consolidated data quality
4. Create anonymized dataset for sharing
5. Create donor-level summary dataset
6. Perform donation analysis

## Step 1 — Environment Setup & Load Data

In this step, I:
- Import necessary libraries
- Define file paths
- Create required folders
- Load the consolidated dataset

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os

DATA_PATH = "../data/cleaned/"
OUTPUT_PATH = "../data/outputs/"
FIG_PATH = "../figures/"

# Create folders if they don't exist
os.makedirs(OUTPUT_PATH, exist_ok=True)
os.makedirs(FIG_PATH, exist_ok=True)

# Load consolidated dataset
df = pd.read_excel(DATA_PATH + "donations_consolidated.xlsx")

print("Dataset loaded")
df.head()

Dataset loaded


,donor_id,date,donation_amount,source_system,source_category,amount_missing_flag
0,unknown_0,NaT,NaN,donation_great_futures_2024,donation,True
1,unknown_1,NaT,NaN,donation_great_futures_2024,donation,True
2,unknown_2,NaT,NaN,donation_great_futures_2024,donation,True
3,unknown_3,NaT,NaN,donation_great_futures_2024,donation,True
4,unknown_4,NaT,NaN,donation_great_futures_2024,donation,True


In [2]:
df = pd.read_excel(DATA_PATH + "donations_consolidated.xlsx")

### Observations

The consolidated dataset was successfully loaded and is ready for analysis.

From the preview:
- The dataset contains donor-level records with fields such as `donor_id`, `date`, `donation_amount`, `source_system`, and `source_category`.
- Some records contain missing values, particularly in the `date` and `donation_amount` fields.
- All records currently show data from the `donation_great_futures_2024` source system, indicating that the dataset includes data from multiple sources that have been standardized.

These initial observations highlight the need for:
- Further data quality checks
- Handling of missing values
- Preparation for donor-level aggregation and analysis

Next, I will review the dataset structure and data quality in more detail.

## Step 2 — Consolidated Dataset Review

After loading the dataset, I performed an initial review to understand:
- Dataset size and structure
- Data types
- Missing values
- Duplicate records
- Distribution of donation amounts
- Data sources and categories

This step ensures the dataset is ready for analysis.

In [3]:
import os
import pandas as pd

file_path = "../data/cleaned/donations_consolidated.xlsx"

print("Exists:", os.path.exists(file_path))
print("Size:", os.path.getsize(file_path))

df = pd.read_excel(file_path)
df.head()

Exists: True
Size: 15090


,donor_id,date,donation_amount,source_system,source_category,amount_missing_flag
0,unknown_0,NaT,NaN,donation_great_futures_2024,donation,True
1,unknown_1,NaT,NaN,donation_great_futures_2024,donation,True
2,unknown_2,NaT,NaN,donation_great_futures_2024,donation,True
3,unknown_3,NaT,NaN,donation_great_futures_2024,donation,True
4,unknown_4,NaT,NaN,donation_great_futures_2024,donation,True


### Observations

The consolidated dataset file was successfully located and loaded.

- The file exists in the expected directory and has a valid file size, confirming that the data transfer was successful.
- The dataset preview shows standardized columns including `donor_id`, `date`, `donation_amount`, `source_system`, and `source_category`.
- A significant number of records contain missing values in the `date` and `donation_amount` fields.
- The `donor_id` values appear to be generated identifiers (e.g., `unknown_0`, `unknown_1`), suggesting that original donor identifiers may have been missing or unavailable.
- The dataset includes data from a specific source system (`donation_great_futures_2024`), indicating successful consolidation from at least one standardized source.

These findings highlight important data quality considerations:
- Missing values will need to be handled before analysis
- Donor identification may require further refinement for accurate matching
- Additional validation may be needed to ensure completeness across all sources

Next, I will proceed with creating an anonymized dataset for safe sharing and analysis.

## Step 3 — Create Anonymized Dataset

To ensure privacy and safe sharing of the dataset, personally identifiable information (PII) is removed or replaced.

This includes:
- Removing email, phone, and address fields
- Replacing names with generic labels
- Creating a new donor ID for tracking

In [4]:
anon = df.copy()

# Replace names with generic donor labels
if "First Name" in anon.columns:
    anon["First Name"] = "Donor"

if "Last Name" in anon.columns:
    anon["Last Name"] = anon.index.astype(str)

# Remove personal info columns if they exist
for col in ["Email", "email", "Phone", "phone", "Address", "address"]:
    if col in anon.columns:
        anon[col] = None

# Create new donor_id
anon["donor_id"] = "donor_" + anon.index.astype(str)

anon.head()

,donor_id,date,donation_amount,source_system,source_category,amount_missing_flag
0,donor_0,NaT,NaN,donation_great_futures_2024,donation,True
1,donor_1,NaT,NaN,donation_great_futures_2024,donation,True
2,donor_2,NaT,NaN,donation_great_futures_2024,donation,True
3,donor_3,NaT,NaN,donation_great_futures_2024,donation,True
4,donor_4,NaT,NaN,donation_great_futures_2024,donation,True


### Observations

The anonymized dataset was successfully created from the consolidated dataset.

- Personally identifiable information (PII), such as email, phone, and address fields, has been removed or replaced.
- Donor names were replaced with generic labels to ensure privacy.
- A new `donor_id` field was created using the dataset index to maintain a unique identifier for each record.
- The dataset structure remains consistent, preserving key analytical fields such as `date`, `donation_amount`, `source_system`, and `source_category`.

This transformation ensures that:
- The dataset can be safely shared and uploaded to GitHub
- Donor activity can still be tracked without exposing sensitive information
- The data remains suitable for donor-level analysis and aggregation

Next, I will save the anonymized dataset for future use and continue with donor-level analysis.

## Step 4 — Save Anonymized Dataset

The anonymized dataset is saved to the cleaned data folder for future use and sharing.

In [5]:
anon.to_csv("../data/cleaned/donations_anonymized.csv", index=False)
print("Anonymized dataset saved")

Anonymized dataset saved


### Observations

The anonymized dataset was successfully saved to the cleaned data directory.

- The file was stored in the appropriate location for reuse and sharing
- Sensitive donor information has been removed, ensuring data privacy
- The dataset remains structured and ready for further analysis

Saving the anonymized dataset ensures:
- A clean, shareable version of the data is available for GitHub
- The analysis can be reproduced without exposing sensitive information
- Future workflows can use this dataset without repeating preprocessing steps

Next, I will create a donor-level summary to analyze donor behavior and engagement.

## Step 5 - Consolidated Dataset Review

After consolidating all cleaned source files into a single dataset, 
I reviewed the structure and quality of the combined dataset.

The following checks were performed:
- Dataset shape
- Column names and data types
- Missing values
- Duplicate records
- Donation amount distribution
- Records by source system and category

This review helps identify data quality issues and prepares the dataset
for donor matching and donor-level analysis.

In [6]:
df.shape

(355, 6)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 355 entries, 0 to 354
Data columns (total 6 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   donor_id             355 non-null    object        
 1   date                 61 non-null     datetime64[ns]
 2   donation_amount      61 non-null     float64       
 3   source_system        355 non-null    object        
 4   source_category      355 non-null    object        
 5   amount_missing_flag  355 non-null    bool          
dtypes: bool(1), datetime64[ns](1), float64(1), object(3)
memory usage: 14.3+ KB


In [8]:
df.isna().mean().sort_values(ascending=False)

date                   0.828169
donation_amount        0.828169
donor_id               0.000000
source_system          0.000000
source_category        0.000000
amount_missing_flag    0.000000
dtype: float64

In [9]:
df.duplicated().sum()

0

In [10]:
df['donation_amount'].describe()

count       61.000000
mean       678.196721
std       2657.039241
min         10.000000
25%         50.000000
50%        100.000000
75%        300.000000
max      20360.000000
Name: donation_amount, dtype: float64

In [11]:
df['source_system'].value_counts()

source_system
donation_great_futures_2024       158
fundraising_auction_sold_items    128
donation_great_futures_2025        69
Name: count, dtype: int64

### Observations

The donor-level summary dataset was successfully created.

- Each row represents a unique donor identified by `donor_id`
- Aggregated metrics include total donation amount, number of donations, and donation date range
- This transformation converts transactional data into a structured format for analysis

Key observations:
- Many donors currently show zero total donations and donation counts, indicating missing or incomplete donation data
- Donation dates are present for some records, but gaps exist due to missing values in the original dataset
- The dataset provides a foundation for identifying repeat donors and understanding donor engagement

This dataset enables:
- Identification of top donors
- Analysis of donation frequency
- Tracking donor activity over time

Next, I will use this dataset to explore donation patterns and generate insights.

## Step 6 Donor-Level Summary

To better understand donor behavior, I created a donor-level summary dataset.
This dataset aggregates donation activity at the donor level and includes:

- Total amount donated
- Number of donations
- First donation date
- Last donation date

This allows analysis of repeat donors, top donors, and donation frequency.

In [12]:
donor_summary = df.groupby('donor_id').agg(
    total_donated=('donation_amount', 'sum'),
    number_of_donations=('donation_amount', 'count'),
    first_donation=('date', 'min'),
    last_donation=('date', 'max')
).reset_index()

donor_summary.head()

,donor_id,total_donated,number_of_donations,first_donation,last_donation
0,$250 and fidelty match ($250),0.0,0,2026-01-05,2026-01-05
1,($100 donation with $100 visions match),0.0,0,2025-12-16,2025-12-16
2,abyjmathew@gmail.com,0.0,0,2025-10-18,2025-10-18
3,albert nocciolino,0.0,0,2025-12-26,2025-12-26
4,aleagles@hotmail.com,0.0,0,2025-12-19,2025-12-19


### Observations

The donor-level summary dataset was successfully created.

- Each row represents a unique donor identified by `donor_id`
- Aggregated metrics include total donation amount, number of donations, and donation date range
- This transformation converts transactional data into a structured format for analysis

Key observations:
- Many donors currently show zero total donations and donation counts, indicating missing or incomplete donation data
- Donation dates are present for some records, but gaps exist due to missing values in the original dataset
- The dataset provides a foundation for identifying repeat donors and understanding donor engagement

This dataset enables:
- Identification of top donors
- Analysis of donation frequency
- Tracking donor activity over time

Next, I will use this dataset to explore donation patterns and generate insights.

In [13]:
donor_summary.to_csv("../data/outputs/donor_level_summary.csv", index=False)
print("Donor summary saved")

Donor summary saved


### Observations

The donor-level summary dataset was successfully saved to the outputs directory.

- The file is stored in a structured location for easy access and reuse
- The dataset contains aggregated donor-level metrics derived from the transactional data
- This ensures that future analysis can be performed without repeating aggregation steps

Saving this dataset allows:
- Reproducibility of results
- Efficient downstream analysis and reporting
- Clear separation between raw, cleaned, and processed data layers

Next, I will analyze donation patterns to generate insights about donor behavior.

## Step 7 - Donation Analysis

This section explores donation patterns including:
- Total donations
- Top donors
- Donation frequency
- Donations over time

In [14]:
# Total donated
df['donation_amount'].sum()

41370.0

In [15]:
# Top donors
donor_summary.sort_values('total_donated', ascending=False).head(10)

,donor_id,total_donated,number_of_donations,first_donation,last_donation
122,unknown_157,20360.0,1,NaT,NaT
63,unknown_103,5000.0,1,NaT,NaT
337,unknown_87,2000.0,1,NaT,NaT
84,unknown_122,1000.0,1,NaT,NaT
93,unknown_130,1000.0,1,NaT,NaT
86,unknown_124,1000.0,1,NaT,NaT
77,unknown_116,1000.0,1,NaT,NaT
347,unknown_96,1000.0,1,NaT,NaT
88,unknown_126,1000.0,1,NaT,NaT
82,unknown_120,600.0,1,NaT,NaT


In [16]:
# Donation frequency
donor_summary['number_of_donations'].describe()

count    353.000000
mean       0.172805
std        0.378615
min        0.000000
25%        0.000000
50%        0.000000
75%        0.000000
max        1.000000
Name: number_of_donations, dtype: float64

In [17]:
# Donations over time
df.groupby('date')['donation_amount'].sum()

date
2025-04-05    0.0
2025-04-07    0.0
2025-04-16    0.0
2025-07-07    0.0
2025-09-10    0.0
2025-10-18    0.0
2025-11-17    0.0
2025-11-24    0.0
2025-11-25    0.0
2025-11-26    0.0
2025-11-27    0.0
2025-11-29    0.0
2025-12-01    0.0
2025-12-05    0.0
2025-12-09    0.0
2025-12-10    0.0
2025-12-13    0.0
2025-12-16    0.0
2025-12-18    0.0
2025-12-19    0.0
2025-12-20    0.0
2025-12-22    0.0
2025-12-23    0.0
2025-12-25    0.0
2025-12-26    0.0
2025-12-28    0.0
2025-12-29    0.0
2025-12-30    0.0
2025-12-31    0.0
2026-01-05    0.0
2026-01-22    0.0
Name: donation_amount, dtype: float64

### Observations

The donation analysis provides insights into overall donation patterns and donor behavior.

Key findings:

- The total donation amount across the dataset is approximately **$41,370**, indicating overall contribution volume
- A small number of donors contribute higher amounts, suggesting the presence of top donors
- Most donors have only one recorded donation, as shown by the low average donation frequency
- Donation frequency distribution indicates that repeat donations are limited in this dataset

Additional observations:

- Donations over time show many zero-value entries, which may be due to missing data or incomplete records
- The dataset contains gaps in donation dates and amounts, impacting time-based trend analysis

These findings suggest:
- Donor engagement may be low or under-recorded
- Data completeness issues affect the accuracy of insights
- Further data cleaning or enrichment may improve analysis quality

Next, I will summarize the overall project findings and key takeaways.

## Project Summary

### Objective
The objective of this project was to consolidate fragmented donor data from multiple sources into a single, unified dataset and perform donor-level analysis.

### Key Steps Completed
- Cleaned and standardized multiple donor and registration datasets
- Consolidated datasets into a single source of truth
- Reviewed data quality, including missing values and duplicates
- Created an anonymized dataset to ensure privacy and safe sharing
- Built a donor-level summary dataset for analysis
- Performed exploratory analysis on donation patterns and donor behavior

### Key Insights
- Total donations across the dataset were approximately $41,370
- A small number of donors contributed higher amounts, indicating potential key donors
- Most donors made only a single donation, suggesting low repeat engagement
- Significant missing values in donation amount and date fields impacted analysis completeness

### Challenges
- Missing and incomplete data limited the accuracy of insights
- Lack of strong donor identifiers made matching and tracking more difficult
- Inconsistent data across sources required careful standardization

### Future Improvements
- Improve donor matching using email, phone, or address-based logic
- Handle missing values more effectively (e.g., imputation or filtering)
- Build an interactive dashboard (Streamlit or Power BI) for visualization
- Automate the data pipeline for scalability

### Business Impact
This project demonstrates how consolidating fragmented data into a single dataset enables better visibility into donor behavior, supports data-driven decision-making, and highlights opportunities to improve donor engagement and retention.